### PR TITLE
feat(auth): Add documentation for next step in auth

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:* https://github.com/aws-amplify/docs/issues/1892

*Description of changes:*

This PR adds a new section under auth called "Sign in next steps"
Screen Shot 2021-02-12 at 2 55 59 PM
Screen Shot 2021-02-12 at 2 56 10 PM
Screen Shot 2021-02-12 at 2 56 22 PM
Screen Shot 2021-02-12 at 2 56 37 PM

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+The signin apis if it succeeds you will get back an associated value of type `AuthSignInResult`, and you can access the next step through the property `nextStep` of `AuthSignInResult`.
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md:2]:
```suggestion
When called successfully, the signin APIs will return an `AuthSignInResult`. Inspect the `nextStep` property in the result to see if additional signin steps required.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
+
+The signin apis if it succeeds you will get back an associated value of type `AuthSignInResult`, and you can access the next step through the property `nextStep` of `AuthSignInResult`. 
+
+```swift
+func signIn(username: String, password: String) {
+    Amplify.Auth.signIn(username: username, password: password) { result in
+        switch result {
+        case .success(signinResult):
+            print("Next step = ")
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md:9]: This example code is missing a critical part, no?
```suggestion
            print("Next step = \(signInResult.nextStep)")
```
Other thought: does it make sense to `switch` over the various possible next steps, here?

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        case .failure(let error):
+            print("Sign in failed \(error)")
+        }
+    }
+}
+```
+
+The next step property is an enum of type `AuthSignInStep`, based on the value you can perform the following steps:
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md:17]:
```suggestion
The `nextStep` property is of enum type `AuthSignInStep`. Depending on its value, your code should take one of the following actions:
```

File: docs/lib/auth/fragments/native_common/signin_next_steps/common.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+Amplify Auth signin flows can be multi step process based on the configuration that you provided when the category was created. In this case you might need to call different apis to successfully authenticate the user into your app. To identify the next step to follow during a signin flow you can inspect the nextStep parameter in the signin result.
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/native_common/signin_next_steps/common.md:1]:
```suggestion
After a user has finished signup, they can proceed to sign in. Amplify Auth signin flows can be multi step processes. The required steps are determined by the configuration you provided when you ran `amplifiy add auth`. Depending on the configuration, you may need to call various APIs to finish authenticating a user's signin attempt. To identify the next step in a signin flow, inspect the `nextStep` parameter in the signin result.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
+If you get `confirmSignInWithSMSMFACode`, the next step is to verify the user through SMS multi-factor authentication. `AuthCodeDeliveryDetails` will contain information regarding where the code has been send. You should ask the user to provide the code that they received via SMS and invoke the `confirmSignIn` api with the SMS code:
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md:1]:
```suggestion
If the next step is `confirmSignInWithSMSMFACode`, Amplify Auth has sent the user a random code over SMS, and is waiting to find out if the user successfully received it. To handle this step, your app's UI must prompt the user to enter the code. After the user enters the code, your implementation must pass the value to Amplify Auth's `confirmSignIn` API.

Note: the signin result also includes an `AuthCodeDeliveryDetails` member. It includes additional information about the code delivery such as the phone number of the SMS recipient.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/30_confirm_custom_challenge.md [hunk 6]
```diff
@@ -0,0 +1,41 @@
+If you receive `confirmSignInWithCustomChallenge`, the next step is to verify the user through custom authentication challenge. The challenge is based on the lambda trigger you have setup for a [custom sign in flow](~/lib/auth/signin_with_custom_flow.md). You should get the custom challenge answer from the user and invoke the `confirmSignIn` api with the custom challenge answer
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/30_confirm_custom_challenge.md:1]:
```suggestion
If the next step is `confirmSignInWithCustomChallenge`, Amplify Auth is awaiting completion of a custom authentication challenge. The challenge is based on the lambda trigger you setup when you configured a [custom sign in flow](~/lib/auth/signin_with_custom_flow.md). To complete this step, you should prompt the user for the custom challenge answer, and pass the answer to the `confirmSignIn` API.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/40_confirm_new_password.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+If you receive `confirmSignInWithNewPassword`, the next step is to provide a new password for the user. You should ask the user for a new password and invoke `confirmSignIn` api with the new password:
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/40_confirm_new_password.md:1]:
```suggestion
If the next step is `confirmSignInWithNewPassword`, Amplify Auth requires a new password for the user before they can proceed. Prompt the user for a new password and pass it to the `confirmSignIn` API.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/50_reset_password.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
+If you receive `resetPassword`, authentication flow could not proceed without resetting the password. The next step is to invoke `resetPassword` api and follow the reset password flow.
```
Comment (jamesonwilliams) [inline docs/lib/auth/fragments/ios/signin_next_steps/50_reset_password.md:1]: Hm. I didn't realize that we have both `resetPassword` and `confirmSignInWithNewPassword`. How are they different? Can we add a sentence or two comparing/contrasting them, to each of the two sections?

**Review by jamesonwilliams — COMMENTED** [review]
Great start! Some alternate phrasings suggested -- feel free to take/ignore as you like.

Looks like the code samples haven't been updated for each step yet? Maybe they ought to show handling a `case` and calling out to some helper function / `confirmSignIn(...)` flow, in response?

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md (hunk 1 above)
Comment (palpatim) [reply docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md:2]: minor grammar fix: "to see if additional signin steps are required"

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md (hunk 5 above)
Comment (palpatim) [reply docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md:1]: "such as the partial phone number of the SMS recipient" (I believe this only includes the last few digits of the number)

File: docs/lib/auth/fragments/ios/signin_next_steps/30_confirm_custom_challenge.md (hunk 6 above)
Comment (palpatim) [reply docs/lib/auth/fragments/ios/signin_next_steps/30_confirm_custom_challenge.md:1]: s/lambda/Lambda/

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+<amplify-block-switcher>
+
+<amplify-block name="Listener (iOS 11+)">
+
+```swift
+func confirmSignIn() {
+    Amplify.Auth.confirmSignIn(challengeResponse: "<confirmation code received via SMS>") { result in
```
Comment (palpatim) [inline docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md:9]: Here and throughout in other code samples: you can make the usage a bit clearer:
```suggestion
func confirmSignIn(confirmationCodeFromUser: String) {
    Amplify.Auth.confirmSignIn(challengeResponse: confirmationCodeFromUser) { result in
```

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        switch result {
+        case .success(let signInResult):
+            print("Confirm sign in succeeded. Next step: \(signInResult.nextStep)")
```
Comment (palpatim) [inline docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md:12]: Here and throughout in other code samples:

```suggestion
            if signInResult.isSignedIn {
                print("Confirm sign in succeeded. The user is now signed in.")
            } else {
                print("Confirm sign in succeeded. Next step: \(signInResult.nextStep)")
                // Switch on the next step to take appropriate actions. If `signInResult.isSignedIn`
                // is true, the next step is 'done', and the user is now signed in.
            }
```

Please confirm that `done` and `isSignedIn` are the correct values.

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+func confirmSignIn() -> AnyCancellable {
+    Amplify.Auth.confirmSignIn(challengeResponse: "<confirmation code received via SMS>")
+        .resultPublisher
+        .sink {
+            if case let .failure(authError) = $0 {
+                print("Confirm sign in failed \(authError)")
+            }
+        }
+        receiveValue: { signInResult in
+            print("Confirm sign in succeeded. Next step: \(signInResult.nextStep)")
```
Comment (palpatim) [inline docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md:34]: Same notes as above re: function argument & success block handling

File: docs/lib/auth/fragments/ios/signin_next_steps/60_confirm_signup.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
+If you receive `confirmSignUp`, authentication flow could not proceed without confirming the user as part of signup. The next step is to invoke `confirmSignUp` api and follow the confirm signup flow.
```
Comment (palpatim) [inline docs/lib/auth/fragments/ios/signin_next_steps/60_confirm_signup.md:1]:
```suggestion
If you receive `confirmSignUp` as a next step, sign up could not proceed without confirming user information such as email or phone number. The next step is to invoke the `confirmSignUp` API and follow the confirm signup flow.
```

File: docs/lib/auth/fragments/ios/signin_next_steps/70_done.md [hunk 13]
```diff
@@ -0,0 +1 @@
+Signin flow is complete when you get `done` and the user is successfully authenticated into the app.
```
Comment (palpatim) [inline docs/lib/auth/fragments/ios/signin_next_steps/70_done.md:1]:
```suggestion
Signin flow is complete when you get `done`. This means the user is successfully authenticated. As a convenience, the SignInResult also provides the `isSignedIn` property, which will be true if the next step is `done`.
```

Please confirm that `isSignedIn` is correct

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md (hunk 2 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md:9]: Changed to switch statement...

File: docs/lib/auth/fragments/ios/signin_next_steps/70_done.md (hunk 13 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/70_done.md]: Added

File: docs/lib/auth/fragments/ios/signin_next_steps/60_confirm_signup.md (hunk 12 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/60_confirm_signup.md]: Updated

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md (hunk 9 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md]: Updated

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md (hunk 11 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md]: Updated

File: docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md (hunk 10 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/20_confirm_sms_mfa.md]: Updated

File: docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md (hunk 3 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/10_signin.md]: Updated

File: docs/lib/auth/fragments/ios/signin_next_steps/50_reset_password.md (hunk 8 above)
Comment (royjit) [reply docs/lib/auth/fragments/ios/signin_next_steps/50_reset_password.md:1]: Added a note in both

**Review by palpatim — APPROVED** [review]

**Pull request merged**